### PR TITLE
GHE auto detect

### PR DESCRIPTION
**Pull request by diasjorge**
Add support to automatically configure ghub to work with the repo's github setup.

### CODE REVIEW

**Comment by diasjorge** [comment]
Only drawback I've seen from enabling the hook is that if you open a project and then go to a magit buffer that was already open, since the magit-status-mode-hook is not run, the ghub-url would remain that of the previous project.

I tried setting the ghub-base-url as buffer local but then the buffer to open pull requests for instance would not have the correct value.

File: magithub-ghe.el [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+;;; Commentary:
+
+;; Integrate with github enterprise
+
+;;; Code:
+
+(require 'magithub-core)
+
+(defcustom magithub-ghe-integration nil
```
Comment (vermiculus) [inline magithub-ghe.el]: I'm not sure this defcustom makes sense – when would you *not* want enterprise support?

File: magithub-ghe.el [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  "Setup Github Enterprise integration."
+  :type 'boolean
+  :group 'magithub)
+
+(defun magithub-auto-detect-github-api-url ()
+  "Set ghub-base-url based on the domain of the repository."
+  (if (magithub-github-repository-p)
+      (let* ((remote-url (magit-get "remote" (magithub-source--remote) "url"))
+             (domain (magithub--url->domain remote-url))
+             (api-url "https://api.github.com"))
```
Comment (vermiculus) [inline magithub-ghe.el]: Why is this separate symbol needed? Why not just `setq-local ghub-base-url` as needed?

File: magithub-core.el [hunk 3]
```diff
@@ -482,6 +482,10 @@ URL may be of several different formats:
   "Tries to parse a remote url into a GitHub repository object"
   (cdr (assq 'sparse-repo (magithub--parse-url url))))
 
+(defun magithub--url->domain (url)
```
Comment (vermiculus) [inline magithub-core.el]: I don't think this is worth a new function; you can use `let-alist` below when you want `.domain`.

Comment (diasjorge) [inline magithub-core.el]: I was trying to keep symmetry with the ->repo function in this case

**Comment by diasjorge** [comment]
Hi, I saw the PR for ghub. For right now this is what I'm using
```
(use-package magithub
  :config
  (magithub-feature-autoinject t)
  (add-hook 'magit-status-mode-hook '(lambda ()
                                       (if (magithub-github-repository-p)
                                           (let-alist (magithub--parse-url (magit-get "remote" (magithub-source--remote) "url"))
                                             (if (string-equal "github.com" .domain)
                                                 (setq ghub-base-url "https://api.github.com")
                                               (setq ghub-base-url (concat "https://" .domain "/api/v3"))))))))
```

It's just good enough for me thanks to the other PR you merged. Maybe when ghub makes those changes I can help you test for GHE support. Cheers!

**Comment by vermiculus** [comment]
> I can help you test for GHE support

I'll hold you to that 😉

**Comment by diasjorge** [comment]
Definitely :) 👍

**Comment by whirm** [comment]
I can help with that too

**Comment by vermiculus** [comment]
I've added you both to [this wiki page](https://github.com/vermiculus/magithub/wiki/GitHub-Enterprise) 😄

**Pull request closed without merging**